### ISSUE TITLE
XSS False negative due to wrong context extraction

### ISSUE THREAD

**Issue by tmendo** [zaproxy/zaproxy#6727]
**Describe the bug**

The TLDR of the bug is that HtmlContextAnalyser.java `getHtmlContexts` function has a bug. In the XSS test I mention below, the rule injects a perfectly constructed `img` tag inside a `div`, which is a working XSS. The problem happens when the threshold is set to High, thus requiring the echo location to match the HTML context we are injecting.

The function `getHtmlContexts`, namely this code at HtmlContextAnalyser.java:179

```
178: // Work out the location in the DOM
179: Element element = src.getEnclosingElement(context.getStart());
```

will report that the enclosing element of the payload is the `img` tag, which is correct with regard to the `getEnclosingElement` function. However, at line 216 that will cause the comparison to fail because the attack context is the `div` and it doesn't match the context of the payload, the `img`.

```
216: } else if (targetContext.matches(context, ignoreFlags)) {
217:                 // Matches the supplied context
218:                 contexts.add(context);
219:             }
220:             
```

So the problem is that the (tag) payload we are adding is adding a new context and it's causing the search for the enclosing element to return one different than expected.
Because the threshold is high, and no contexts are returned, so the rule fails to add to alert the XSS, despite being a valid XSS matching the desired condition for the High threshold.

HTML response with the working payload. The payload is `<img src=x onerror=alert(1);>`

```
<html>
	<table>
		<tr>
			<td valign='top'>
			  <div id='0'>
				<img src=x onerror=alert(1);>
			  </div>
			</td>
		</tr>
	</table>
</html>
```

**To Reproduce**
1. Apply the following changes, and run the test `shouldReportXssInsideDivWithFilteredScript`

```
index cc4f0fc30..a2b7b466d 100644
--- a/addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java
+++ b/addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java
@@ -1024,6 +1024,7 @@ class CrossSiteScriptingScanRuleUnitTest extends ActiveScannerTest<CrossSiteScri
 
         HttpMessage msg = this.getHttpMessage(test + "?name=test");
 
+        this.rule.setAlertThreshold(AlertThreshold.HIGH);
         this.rule.init(msg, this.parent);
 
         this.rule.scan();
```
2. notice the test fails because no alert is reported

**Expected behavior**
The test should pass, because the payload is reflected in the context it is injected.

**Software versions**
 - ZAP: latest, stable
 - Add-on: ascanrules release, version 41
commit `43478c734b0edc7ac4460a2e958d4cb67bbc96d2`, main branch

**Would you like to help fix this issue?**
I would, but I'm afraid of the implications of changing such an inner function of ZAP that does all the parsing and context analysis, and what could break in any other extension. I can implement the changes but some guidance would be helpfull.

**Comment by ArkaprabhaChakraborty** [comment]
Uhh, can I work on this?

**Comment by thc202** [comment]
Sure.

**Comment by psiinon** [comment]
Just confirmed with @ArkaprabhaChakraborty that they want to be assigned #7157 so this issue is still available for anyone else to take on.

**Comment by tmendo** [comment]
@DiogoMRSilva and I have fix this issue (and other XSS ones) and will try to submit a PR this week (sorry for being late to this)

**Comment by tmendo** [comment]
fixed by https://github.com/zaproxy/zap-extensions/commit/06ced036bb50fabdf1f6601092f5025604cce559
